### ISSUE TITLE
Support C++-defined methods on shared structs

### ISSUE THREAD

**Issue by dtolnay** [dtolnay/cxx#376]
```rust
// src/main.rs

#[cxx::bridge]
mod ffi {
    struct Address {
        house_number: u64,
        street: String,
    }

    extern "C" {
        fn print(self: &Address);
    }
}

fn main() {
    let addr = ffi::Address {
        house_number: 1060,
        street: "W Addison St".to_owned(),
    };
    addr.print();
}
```

```cpp
// src/address.cc

#include "repro/src/main.rs.h"

void Address::print() const {
  /*...*/
}
```

Currently this generates noncompilable C++ code because we aren't putting a `print` member function into the generated `Address` class.

```console
target/debug/build/repro/out/cxxbridge/sources/repro/src/main.rs.cc: In function ‘void cxxbridge05$Address$print(const Address&)’:
target/debug/build/repro/out/cxxbridge/sources/repro/src/main.rs.cc:54:47: error: ‘print’ is not a member of ‘Address’
   54 |   void (Address::*print$)() const = &Address::print;
      |                                               ^~~~~
```

```cpp
// generated
struct Address final {
  uint64_t house_number;
  ::rust::String street;

  // needs to include:
  // void print() const;
};
```

**Comment by adetaylor** [comment]
Sorry, I did not mean to close this! I meant, in fact, to ask a question but realized I needed to do some research first. Nothing to see here. Move along.

**Comment by 2xsaiko** [comment]
This also happens for the same thing but on the Rust side:
```rust
#[cxx::bridge(namespace = "mcrtlib::ffi")]
mod ffi {
    pub struct DataSource {
        pub inner: Box<DataSource_>,
    }

    extern "Rust" {
        type DataSource_;

        fn create_dir(self: &DataSource, path: &str) -> Result<()>;
    }
}

impl ffi::DataSource {
    fn create_dir(&self, path: &str) -> Result<(), datasource::Error> {
        self.inner.create_dir(path)
    }
}
```
gives
```
/home/saiko/src/_mc/mcrestool/cmake-build-debug/library/lib.rs.cpp:508:6: error: no declaration matches ‘void mcrtlib::ffi::DataSource::create_dir(rust::cxxbridge05::Str) const’
  508 | void DataSource::create_dir(::rust::Str path) const {
      |      ^~~~~~~~~~
/home/saiko/src/_mc/mcrestool/cmake-build-debug/library/lib.rs.cpp:508:6: note: no functions named ‘void mcrtlib::ffi::DataSource::create_dir(rust::cxxbridge05::Str) const’
/home/saiko/src/_mc/mcrestool/cmake-build-debug/library/lib.rs.cpp:467:8: note: ‘struct mcrtlib::ffi::DataSource’ defined here
  467 | struct DataSource final {
      |        ^~~~~~~~~~
```
generated:
```cpp
struct DataSource final {
  ::rust::Box<::mcrtlib::ffi::DataSource_> inner;
};
```

It only works if you define it on the opaque type, `DataSource_` in this case. Which means unless I'm missing something you can't have a type that you can both use from C++ without a Box around it and directly call methods on.

**Comment by dtolnay** [comment]
Fixed in #437. @2xsaiko could you try it with cxx 0.5.6?
